### PR TITLE
Fix handling of empty JSON arrays during sync

### PR DESCRIPTION
**Pull request by mwest1066**
Fixees #2062

The SQL expression that we were using to convert JSON arrays to SQL
arrays was converting zero-length arrays to NULL, rather than to
zero-length SQL arrays. This PR adds explicit PLPGSQL functions to do
this conversion. These map:
```
NULL::jsonb         -> NULL::text[]
'null'::jsonb       -> NULL::text[]
'[]'::jsonb         -> array[]::text[]
'["a"]'::jsonb      -> array['a']::text[]
'["a", "b"]'::jsonb -> array['a', 'b']::text[]
```

### CODE REVIEW

File: sprocs/jsonb_array_to_double_precision_array.sql [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    END IF;
+
+    len := jsonb_array_length(jsonb_array);
+
+    IF len IS NULL THEN
+        double_precision_array := NULL;
+    ELSIF len = 0 THEN
+        double_precision_array := array[]::double precision[];
+    ELSE
+        double_precision_array := (SELECT array_agg(v::double precision) FROM jsonb_array_elements_text(jsonb_array) AS tmp(v));
```
Comment (tbretl) [inline sprocs/jsonb_array_to_double_precision_array.sql]: The reason for `AS tmp(v)` is not obvious to me here.

File: sprocs/jsonb_array_to_text_array.sql [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    END IF;
+
+    len := jsonb_array_length(jsonb_array);
+
+    IF len IS NULL THEN
+        text_array := NULL;
+    ELSIF len = 0 THEN
+        text_array := array[]::text[];
+    ELSE
+        text_array := (SELECT array_agg(v) FROM jsonb_array_elements_text(jsonb_array) AS tmp(v));
```
Comment (tbretl) [inline sprocs/jsonb_array_to_text_array.sql]: Again, I don't understand the `AS tmp(v)` syntax.

File: tests/testSproc-check_assessment_access.js [hunk 3]
```diff
@@ -161,6 +161,42 @@ describe('sproc check_assessment_access* tests', function() {
             });
         });
 
+        it('pass if all parameter match match except UID with no uids', function(callback) {
```
Comment (tbretl) [inline tests/testSproc-check_assessment_access.js]: I think you mean `pass if all parameters match except UID with no uids`

File: sprocs/jsonb_array_to_double_precision_array.sql (hunk 1 above)
Comment (mwest1066) [reply sprocs/jsonb_array_to_double_precision_array.sql]: It's a table alias, so I don't have to write
```
array_agg(jsonb_array_elements_text::double precision)
```

The general table alias syntax is `AS tablename(col1name, col2name)`. You don't have to provide the column aliases if you don't want them. In this case they aren't strictly necessary, so I just pushed a change `tmp(v) -> v`.

File: tests/testSproc-check_assessment_access.js (hunk 3 above)
Comment (mwest1066) [reply tests/testSproc-check_assessment_access.js]: fixed

**Pull request merged**